harness/determined/profiler.py

```diff
@@ -0,0 +1,524 @@
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_IOPS_METRIC, iops_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_THRU_READ_METRIC, disk_read_thru_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.DISK_THRU_WRITE_METRIC, disk_write_thru_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.FREE_MEM_METRIC, free_memory_measurement)
+                self.current_batch.add_nongpu_measurement(SysMetricType.SIMPLE_CPU_UTIL_METRIC, cpu_util_measurement)
+
+                last_measurement_time = time.time()
+
+
+            # Check if it is time to flush the batch and start a new batch
+            if time.time() - batch_start_time > self.FLUSH_INTERVAL:
+                self.enqueue_for_async_send(self.current_batch.convert_to_post_format())
+                self.current_batch.clear()
+                batch_start_time = time.time()
+
+            time.sleep(0.02)
+
+
+
+
+
+
+
+
+
+# This is a thread that exists solely so that we can make API calls without blocking
+# It has a Queue through which work is sent to the thread
+class ProfilerSenderThread(threading.Thread):
+    POLL_INTERVAL_SECS = 0.5
+    def __init__(self, inbound_queue: queue.Queue, master_url: str) -> None:
+        self.master_url = master_url
+        self.inbound_queue = inbound_queue
+        self.quitting = False
+        super().__init__()
+
+    def run(self) -> None:
+        while True:
+            if self.quitting:
+                break
+
+            try:
+                batch_to_send = self.inbound_queue.get_nowait()
+            except queue.Empty:
+                time.sleep(ProfilerSenderThread.POLL_INTERVAL_SECS)
+                continue
+
+            self.send_batch(batch_to_send)
+
+    # This is a blocking operation that must handle all exceptions gracefully
+    def send_batch(self, post_bodies):
+        # TODO: In the case of repeated failure, this can take a long time
+        #       and we have no mechanism to handle the inbound queue filling up
+        max_attempts = 5
+
+        # TODO: Convert to work with new multiple batch API
+        for post_body in post_bodies:
+            for i in range(max_attempts):
+                try:
+                    api.post_trial_profiler_metrics(
+                        self.master_url,
+                        post_body["values"],
+                        post_body["batches"],
+                        post_body["timestamps"],
+                        post_body["labels"],
+                    )
+                    break
+                except Exception as e:
+                    # TODO: We could handle specific API errors differently as some are non-recoverable
+                    # TODO: Log info about error
+                    time.sleep(i**2)  # exponential backoff
```

Comment:
We might end up getting pretty backed up here (this could wait 31 seconds per batch total while they're piling in if the master blips). It's probably ok to try one time in a constant interval and bail if we can't. It's just a single batch.